pkg/storage/ingest/pusher.go

```diff
@@ -241,163 +253,239 @@ func (c pusherConsumer) shouldLogClientError(ctx context.Context, err error) (bo
+func newParallelStorageShards(numTimeSeriesPerFlush prometheus.Histogram, numShards int, batchSize int, capacity int, pusher Pusher) *parallelStorageShards {
+	p := &parallelStorageShards{
 		numShards:             numShards,
-		upstream:              upstream,
+		pusher:                pusher,
+		capacity:              capacity,
 		numTimeSeriesPerFlush: numTimeSeriesPerFlush,
 		batchSize:             batchSize,
 		wg:                    &sync.WaitGroup{},
 		errs:                  make(chan error, numShards),
-		unfilledShards:        make([]flushableWriteRequest, numShards),
-	}
-	shards := make([]chan flushableWriteRequest, numShards)
-	pusher.wg.Add(numShards)
-	for i := range shards {
-		pusher.unfilledShards[i].WriteRequest = &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()}
-		shards[i] = make(chan flushableWriteRequest, buffer)
-		go pusher.runShard(shards[i])
 	}
-	go func() {
-		pusher.wg.Wait()
-		close(pusher.errs)
-	}()
 
-	pusher.shards = shards
-	return pusher
+	p.start()
+
+	return p
 }
 
-func (p *shardingPusher) PushToStorage(ctx context.Context, request *mimirpb.WriteRequest) error {
+func (p *parallelStorageShards) ShardWriteRequest(ctx context.Context, request *mimirpb.WriteRequest) error {
 	var (
 		builder         labels.ScratchBuilder
 		nonCopiedLabels labels.Labels
 		errs            multierror.MultiError
+		wg              sync.WaitGroup
 	)
+
+	// Collect the errors in a separate goroutine so that we try not to block whilst processing.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for err := range p.errs {
+			errs.Add(err)
+		}
+	}()
+
 	for _, ts := range request.Timeseries {
 		mimirpb.FromLabelAdaptersOverwriteLabels(&builder, ts.Labels, &nonCopiedLabels)
 		shard := nonCopiedLabels.Hash() % uint64(p.numShards)
 
-		s := p.unfilledShards[shard]
-		// TODO dimitarvdimitrov support metadata and the rest of the fields; perhaps cut a new request for different values of SkipLabelNameValidation?
-		s.Timeseries = append(s.Timeseries, ts)
-		s.Context = ctx // retain the last context in case we have to flush it when closing shardingPusher
+		// TODO: Add metrics to measure how long are items sitting in the queub efore they are flushed.
+		p.shards[shard].AddToBatch(ctx, ts)
+	}
 
-		if len(s.Timeseries) < p.batchSize {
-			continue
-		}
-		p.unfilledShards[shard] = flushableWriteRequest{
-			WriteRequest: &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()},
-		}
+	// Close the channel so that the goroutine collecting errors can finish.
+	close(p.errs)
 
-	tryPush:
-		for {
-			select {
-			case p.shards[shard] <- s:
-				break tryPush
-			case err := <-p.errs:
-				// we might have to first unblock the shard before we can flush to it
-				errs.Add(err)
-			}
-		}
+	wg.Wait()
 
-		// drain errors to avoid blocking the shard loop
-	drainErrs:
-		for {
-			select {
-			case err := <-p.errs:
-				errs.Add(err)
-			default:
-				break drainErrs
-			}
-		}
```

Comment:
In the end, I realised the value of these two selects - the trade-off here is that I have to `p.errs = make(chan error, p.numShards)` re-create the error channels once we're done as we use the closing of an channels as an indication that no more data was mean to go through. I'm open to the idea of bring this back if you feel strongly about it as I'm not sure how much it'll cost in terms of resources us to re-create the channels again.